bin/substrate/src/main.rs

```diff
@@ -0,0 +1,208 @@
+mod key_server;
+mod key_server_set;
+mod runtime;
+mod service;
+mod substrate_client;
+mod transaction_pool;
+
+use std::{
+	io::Write,
+	sync::Arc,
+};
+use futures::{FutureExt, TryFutureExt};
+use log::error;
+use parity_crypto::publickey::{KeyPair, public_to_address};
+use primitives::{
+	executor::{TokioRuntime, tokio_runtime},
+	key_server_key_pair::InMemoryKeyServerKeyPair,
+};
+
+fn main() {
+	initialize();
+
+	let arguments = match arguments::parse_arguments(None) {
+		Ok(arguments) => arguments,
+		Err(error) => {
+			error!(
+				target: "secretstore",
+				"Failed to parse arguments: {:?}",
+				error,
+			);
+
+			return;
+		}
+	};
+
+	let _ = futures::executor::LocalPool::new()
+		.run_until(
+			run_key_server(arguments)
+				.map_err(|error| {
+					error!(
+						target: "secretstore",
+						"Failed to start: {:?}",
+						error,
+					);
+				})
+		);
+}
+
+/// Run key server ad blockchain service.
+async fn run_key_server(arguments: arguments::Arguments) -> Result<(), String> {
+	// we still need tokio 0.1 runtime to run SS :/
+	let tokio_runtime = tokio_runtime()
+		.map_err(|err| format!("Error creating tokio runtime: {}", err))?;
+	// and since not everything in SS is async, we need an additional
+	// futures executor that we'll use to run futures in sync functions
+	let thread_pool = futures::executor::ThreadPool::new()
+		.map_err(|err| format!("Error creating thread pool: {}", err))?;
+
+	// start key server and services
+	let (client, key_server_set, best_sender) = start_key_server(
+		arguments,
+		&tokio_runtime,
+		thread_pool,
+	).await?;
+
+	let mut fut_finalized_headers = client.subscribe_finalized_heads().await
+		.map_err(|err| format!("Failed to subscribe to finalized blocks: {:?}", err))?;
+
+	loop {
+		futures::select! {
+			finalized_header = fut_finalized_headers.next().fuse() => {
+				let finalized_block = (finalized_header.number, finalized_header.hash());
+				client.set_best_block(finalized_block);
+				key_server_set.set_best_block(finalized_block);
+				if let Err(error) = best_sender.unbounded_send(finalized_block.1) {
+					error!(
+						target: "secretstore",
+						"Failed to send finalized block: {:?}",
+						error,
+					);
+				}
+			},
+		}
+	}
+}
+
+/// Start key server and blockchain service.
+async fn start_key_server(
+	arguments: arguments::Arguments,
+	tokio_runtime: &TokioRuntime,
+	thread_pool: futures::executor::ThreadPool,
+) -> Result<(
+	substrate_client::Client,
+	Arc<key_server_set::OnChainKeyServerSet>,
+	futures::channel::mpsc::UnboundedSender<runtime::BlockHash>,
+), String> {
+	// let's connect to Substrate node first
+	let client = substrate_client::Client::new(
+		&format!("ws://{}:{}", arguments.sub_host, arguments.sub_port),
+		runtime::create_transaction_signer(
+			&arguments.sub_signer,
+			arguments.sub_signer_password.as_deref(),
+		)?,
+	).await.map_err(|error| format!("Failed to start substrate client: {:?}", error))?;
+
+	// TODO: use db key storage
```

Comment:
This is a next thing to fix - I'm preparing a PR that will reintroduce persistent keys storage and then will make a PR that fixes this TODO.